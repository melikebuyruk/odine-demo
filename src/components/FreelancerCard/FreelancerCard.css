.freelancer-card {
    max-width: 345px;
    border: 1px solid #e0e0e0;
    border-radius: 8px;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
    transition: transform 0.3s ease, box-shadow 0.3s ease;
}

.freelancer-card:hover {
    transform: translateY(-5px);
    box-shadow: 0 6px 12px rgba(0, 0, 0, 0.2);
}

.card-media {
    height: 194px;
    border-bottom: 1px solid #e0e0e0;
}

.card-actions {
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.expand-more {
    margin-left: auto;
    transition: transform 0.3s;
}

.expand-more[aria-expanded='true'] {
    transform: rotate(180deg);
}

.card-content {
    padding: 16px;
}

.card-content h5 {
    font-size: 1.25rem;
    font-weight: 600;
}

.card-details {
    padding: 16px;
    background-color: #f9f9f9;
    border-top: 1px solid #e0e0e0;
}