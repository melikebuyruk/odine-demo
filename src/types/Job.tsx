export interface Job {
  id: number;
  userId: number;
  title: string; 
  body: string; 
  comments: Comment[];
}

export interface Comment {
  id: number; 
  body: string; 
}
