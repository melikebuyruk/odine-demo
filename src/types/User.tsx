export interface User {
    
  id: number;
  name: string;
  username: string;
  email: string;
  phone: string;
  website: string;
  photo: string;
  jobCount: number;
  address: {
    street: string;
    suite?: string;
    city: string;
    zipcode?: string;
  };

  company: {
    name: string;
    catchPhrase: string;
    bs: string;
  }
}
