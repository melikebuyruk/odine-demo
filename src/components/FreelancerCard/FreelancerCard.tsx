import * as React from 'react';
import { styled } from '@mui/material/styles';
import Card from '@mui/material/Card';
import CardMedia from '@mui/material/CardMedia';
import CardContent from '@mui/material/CardContent';
import CardActions from '@mui/material/CardActions';
import Collapse from '@mui/material/Collapse';
import IconButton, { IconButtonProps } from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import TurnedInIcon from '@mui/icons-material/TurnedIn';
import TurnedInNotIcon from '@mui/icons-material/TurnedInNot';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import { User } from '../../types/User';
import Portfolio from '../Portfolio/Portfolio';
import HireFreelancer from '../HireFreelancer/HireFreelancer';
import './FreelancerCard.css';

interface ExpandMoreProps extends IconButtonProps {
  expand: boolean;
}

interface FreelancerCardProps {
  User: User;
  onSave: () => void;
  isSaved: boolean;
  photo:string;
}

const ExpandMore = styled((props: ExpandMoreProps) => {
  const { expand, ...other } = props;
  return <IconButton {...other} />;
})(({ theme, expand }) => ({
  marginLeft: 'auto',
  transition: theme.transitions.create('transform', {
    duration: theme.transitions.duration.shortest,
  }),
  transform: expand ? 'rotate(180deg)' : 'rotate(0deg)',
}));

const FreelancerCard: React.FC<FreelancerCardProps> = ({ User, onSave, isSaved,photo}) => {
  const [expanded, setExpanded] = React.useState(false);


  const handleExpandClick = () => {
    setExpanded(!expanded);
  };

  return (
    <Card className="freelancer-card">
      <CardMedia className="card-media" component="img" height="194" image={photo} alt={User.name} />
      <CardContent>
        <Typography variant="h5">{User.name}</Typography>
      </CardContent>
      <Portfolio User={User} />
      <CardActions disableSpacing>
        <HireFreelancer User={User} />
        <IconButton onClick={onSave} aria-label="Save" data-testid={isSaved ? "save-button-saved" : "save-button-unsaved"}>
  {isSaved ? (
    <TurnedInIcon data-testid="TurnedInIcon" color="primary" />
  ) : (
    <TurnedInNotIcon data-testid="TurnedInNotIcon" />
  )}
</IconButton>


        <ExpandMore
          expand={expanded}
          onClick={handleExpandClick}
          aria-expanded={expanded}
          aria-label="show more"
        >
          <ExpandMoreIcon />
        </ExpandMore>
      </CardActions>
      <Collapse in={expanded} timeout="auto" unmountOnExit>
        <CardContent>
          <Typography paragraph>Email: {User.email}</Typography>
          <Typography paragraph>Phone: {User.phone}</Typography>
          <Typography paragraph>Finished Job Count: {User.jobCount}</Typography>
          <Typography paragraph>City: {User.address.city}</Typography>
        </CardContent>
      </Collapse>
    </Card>
  );
};

export default FreelancerCard;
