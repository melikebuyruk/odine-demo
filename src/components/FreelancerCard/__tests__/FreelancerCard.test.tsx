import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import FreelancerCard from '../FreelancerCard';
import { User } from '../../../types/User';

const mockUser: User = {
  id: 1,
  name: 'Melike Buyruk',
  username: 'mbuyruk',
  email: 'melikke.buyruk@gmail.com',
  phone: '123-456-7890',
  website: 'melike.com',
  photo: 'https://via.placeholder.com/150',
  jobCount: 10,
  address: {
    street: 'Main St',
    suite: 'Apt. 101',
    city: 'New York',
    zipcode: '10001',
  },
  company: {
    name: 'Odine',
    catchPhrase: 'We hire talents!',
    bs: 'tech solutions',
  },
};

const mockOnSave = jest.fn();

describe('FreelancerCard Component', () => {
  beforeEach(() => {
    render(
      <FreelancerCard
        User={mockUser}
        onSave={mockOnSave}
        isSaved={false}
        photo={mockUser.photo}
      />
    );
    jest.clearAllMocks();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <FreelancerCard
        User={mockUser}
        onSave={mockOnSave}
        isSaved={false}
        photo="https://via.placeholder.com/345x194"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should expand and collapse when the expand button is clicked', async () => {
    expect(screen.queryByText(/Email: melikke.buyruk@gmail.com/i)).not.toBeInTheDocument();
    const expandButton = screen.getByLabelText(/show more/i);
    fireEvent.click(expandButton);
    expect(screen.getByText(/Email: melikke.buyruk@gmail.com/i)).toBeInTheDocument();
    fireEvent.click(expandButton);
    await waitFor(() => {
      expect(screen.queryByText(/Email: melikke.buyruk@gmail.com/i)).not.toBeInTheDocument();
    });
  });

  it('should render TurnedInIcon when isSaved is true', () => {
  render(
    <FreelancerCard
      User={mockUser}
      onSave={mockOnSave}
      isSaved={true}
      photo={mockUser.photo}
    />
  );
  const saveButton = screen.getByTestId('save-button-saved');
  expect(saveButton).toContainElement(screen.getByTestId('TurnedInIcon'));
});


});
