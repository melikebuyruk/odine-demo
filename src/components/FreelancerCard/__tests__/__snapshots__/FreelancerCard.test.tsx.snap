// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FreelancerCard Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root freelancer-card css-1j74uew-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <img
      alt="Melike Buyruk"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img card-media css-1kuwa16-MuiCardMedia-root"
      height="194"
      src="https://via.placeholder.com/345x194"
    />
    <div
      class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
      >
        Melike Buyruk
      </h5>
    </div>
    <div>
      <button
        class="css-1t9l90i"
        data-testid="open-portfolio"
        style="background-color: rgb(255, 255, 255); color: rgb(25, 118, 210);"
        type="button"
      >
        Portfolio
      </button>
    </div>
    <div
      class="MuiCardActions-root css-bbbzor-MuiCardActions-root"
    >
      <div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
          data-testid="hire-open-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="HomeRepairServiceIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 16h-2v-1H8v1H6v-1H2v5h20v-5h-4zm2-8h-3V6c0-1.1-.9-2-2-2H9c-1.1 0-2 .9-2 2v2H4c-1.1 0-2 .9-2 2v4h4v-2h2v2h8v-2h2v2h4v-4c0-1.1-.9-2-2-2m-5 0H9V6h6z"
            />
          </svg>
        </button>
      </div>
      <button
        aria-label="Save"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
        data-testid="save-button-unsaved"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="TurnedInNotIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2m0 15-5-2.18L7 18V5h10z"
          />
        </svg>
      </button>
      <button
        aria-expanded="false"
        aria-label="show more"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-82czz1-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
